xla/service/elemental_ir_emitter.cc

```diff
@@ -534,6 +534,572 @@ llvm::Value* EmitF8e4m3b11fnuzToF16(llvm::Value* f8_value,
+          (dest_mantissa_bits + 1 + primitive_util::ExponentWidth(input_type) +
+           (primitive_util::UnderflowExponent(input_type) -
+            primitive_util::UnderflowExponent(output_type))));
+
+  // Select the last mantissa bit based on whether the input is a normal number.
+  llvm::Value* x_normal_pred = IsNormalNumber(input_type, value, b);
+  x_normal_pred = b->CreateOr(x_normal_pred, IsZero(input_type, value, b));
+
+  return b->CreateSelect(x_normal_pred, shift, zero_count);
+}
+
+// Compute the bias for round-to-nearest-even for the input value. This takes
+// into account whether the input value is a normal number and whether
+// it will map to a normal number in the output type.
+llvm::Value* DynamicBias(PrimitiveType input_type, llvm::Value* x_as_int,
+                         PrimitiveType output_type, llvm::IRBuilder<>* b) {
+  llvm::Type* int_type = b->getIntNTy(primitive_util::BitWidth(input_type));
+
+  // Find the bit position of the last mantissa bit.
+  llvm::Value* shift = LastMantissaBit(input_type, x_as_int, output_type, b);
+
+  // Compute the mask to select that bit.
+  llvm::Value* last_mantissa_bit_mask =
+      b->CreateShl(llvm::ConstantInt::get(int_type, 0x1u), shift);
+
+  // Given the mantissa bit mask, compute the rounding bias bits.
+  llvm::Value* base_rounding_bias = b->CreateLShr(last_mantissa_bit_mask, 0x1u);
+  base_rounding_bias =
+      b->CreateSub(base_rounding_bias, llvm::ConstantInt::get(int_type, 0x1u));
+
+  // Select the last mantissa bit, and shift it down to the lsb.
+  llvm::Value* x_last_mantissa_bit =
+      b->CreateLShr(b->CreateAnd(x_as_int, last_mantissa_bit_mask), shift);
+
+  // Add the last mantissa lsb into the rounding bias.
+  return b->CreateAdd(x_last_mantissa_bit, base_rounding_bias);
+}
+
+// Given an unbiased exponent and mantissa with no implicit bits, returns the
+// mantissa for the output type.
+llvm::Value* BuildOutputMantissa(PrimitiveType input_type,
+                                 llvm::Value* exponent, llvm::Value* mantissa,
+                                 PrimitiveType output_type,
+                                 llvm::IRBuilder<>* b) {
+  llvm::Type* input_int_type =
+      b->getIntNTy(primitive_util::BitWidth(input_type));
+
+  // Count the number of leading zeros, excluding the bits that would contain
+  // the exponent.
+  llvm::Value* zero_count =
+      b->CreateBinaryIntrinsic(llvm::Intrinsic::ctlz, mantissa,
+                               llvm::ConstantInt::get(b->getInt1Ty(), 0x0u));
+  zero_count = b->CreateSub(
+      zero_count,
+      llvm::ConstantInt::get(input_int_type,
+                             primitive_util::ExponentWidth(input_type)));
+
+  // The amount to shift the normal mantissa down.
+  llvm::Value* shift = b->CreateSub(
+      llvm::ConstantInt::get(input_int_type,
+                             primitive_util::SignificandWidth(input_type) -
+                                 primitive_util::SignificandWidth(output_type)),
```

Comment:
To reduce verbosity, add some using-declarations to commonly used `primitive_util` functions at the top of the file, e.g.:

```c++
using primitive_util::SignificandWidth
```

That way you don't have to prefix these functions with `primitive_util::`.